src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs

```diff
@@ -0,0 +1,75 @@
+ */
+
+using System;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using AspNet.Security.OAuth.Extensions;
+using Microsoft.AspNet.Authentication;
+using Microsoft.AspNet.Authentication.OAuth;
+using Microsoft.AspNet.Http.Authentication;
+using Microsoft.AspNet.WebUtilities;
+using Microsoft.Extensions.Internal;
+using Newtonsoft.Json.Linq;
+
+namespace AspNet.Security.OAuth.Instagram {
+    public class InstagramAuthenticationHandler : OAuthHandler<InstagramAuthenticationOptions> {
+        public InstagramAuthenticationHandler([NotNull] HttpClient client)
+            : base(client) {}
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens) {
+            var userInformationEndpoint = QueryHelpers.AddQueryString(Options.UserInformationEndpoint, name: "access_token", value: tokens.AccessToken);
+
+            if (Options.UseSignedRequests) {
+                var signature = SignRequest("/users/self",
+                    tokens.AccessToken, Options.ClientSecret);
+
+                userInformationEndpoint = QueryHelpers.AddQueryString(userInformationEndpoint, name: "sig",
+                    value: signature);
+            }
+
+            var request = new HttpRequestMessage(HttpMethod.Get, userInformationEndpoint);
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+
+            var response =
+                await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+            response.EnsureSuccessStatusCode();
+
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync()).Value<JObject>("data");
+
+            identity.AddOptionalClaim(ClaimTypes.NameIdentifier, InstagramAuthenticationHelper.GetIdentifier(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim(ClaimTypes.Name, InstagramAuthenticationHelper.GetFullName(payload), Options.ClaimsIssuer);
+
+            var context = new OAuthCreatingTicketContext(Context, Options, Backchannel, tokens, payload) {
+                Principal = new ClaimsPrincipal(identity),
+                Properties = properties
+            };
+
+            await Options.Events.CreatingTicket(context);
+
+            if (context.Principal?.Identity == null) {
+                return null;
+            }
+
+            return new AuthenticationTicket(context.Principal, context.Properties, context.Options.AuthenticationScheme);
+        }
+
+        private static string SignRequest([NotNull] string endpoint, [NotNull] string accessToken,
+            string secret) {
+            var token = Encoding.UTF8.GetBytes($"{endpoint}|access_token={accessToken}");
```

Comment:
Hmmmm, the way the signature is formatted is not really foolproof and won't work if the user adds query parameters in the userinfo URL.

I hacked a different version. Could you please give it a try? (I don't have an Instagram account :smile:)

```csharp
protected virtual string ComputeSignature(string address) {
    using (var algorithm = new HMACSHA256(Encoding.UTF8.GetBytes(Options.ClientSecret))) {
        // Extract the parameters from the query string.
        var parameters = from parameter in QueryHelpers.ParseQuery(address)
                         orderby parameter.Key
                         from value in parameter.Value
                         orderby value
                         select $"{parameter.Key}={value}";

        // See https://www.instagram.com/developer/secure-api-requests/
        // for more information about the signature format.
        var bytes = Encoding.UTF8.GetBytes($"/users/self|{string.Join("|", parameters)}");

        // Compute the HMAC256 signature.
        var hash = algorithm.ComputeHash(bytes);

        // Convert the hash to its lowercased hexadecimal representation.
        return BitConverter.ToString(hash).Replace("-", "").ToLower();
    }
}
```